### DISCUSSION TITLE
localising-the-displayattribute-in-asp-net-core-1-1/

### DISCUSSION THREAD

**Discussion by andrewlock** [andrewlock/blog-comments#219] [Announcements]
# localising-the-displayattribute-in-asp-net-core-1-1/

In this post, I recap localisation in ASP.NET Core, and see how localising the  DisplayAttribute changed in ASP.NET Core 1.1. Spoiler alert - it just works!

https://andrewlock.net/localising-the-displayattribute-in-asp-net-core-1-1/

[a comment by an automated account is not shown]

**Comment by mrpmorris** [comment]
As per our discussion on bsky...

I'm not a fan of IStringLocalizer<T>. If the key isn't present then you just get the key back, which means coders can easily get the keys wrong without compiler errors and have parts of the app not localized.

When it comes to ValidatorAttribute, the default error messages come from a pre-defined format string that has a placeholder for the member name. For example, "The {0} field is required". This is why in your example you have to set the ErrorMessage on every validator - but there are some problems with that approach.

Firstly, anyone who uses a validation summary will just see "This field is required" multiple times instead of "The Name field is required." etc.

Secondly, some validation attributes have multiple placeholders in their format strings. For example, StringLength has the message "{0} has a maximum length of {1}." In this case you'd either need to have a single string that doesn't tell the user the actual maximum length "The field is too long", or you'd need a unique string for every unique string length in your app "The field has a maximum length of 8", "The field has a maximum length of 16", and so on.

It would be nice if we could just set ErrorMessageResourceType and a convention for ErrorMessageResourceName at a global level. I have had an email exchange recently with some people at MS about this but it's too early to know if they'll agree or not.

In the meantime, (shameful plug) I wrote this Fody weaver to do exactly that. So, with a single line of code I can have translated validation messages, no errors from repetitive manual work, and they support the field name and property placeholders such as the maximum string length.

https://github.com/mrpmorris/AutoLocalize
